cli/src/semdep/parse_lockfile.py

```diff
@@ -45,38 +51,40 @@ def extract_npm_lockfile_hash(s: str) -> Dict[str, List[str]]:
     return {algorithm: [base64.b16encode(decode_base_64).decode("ascii").lower()]}
 
 
-def parse_Yarnlock_str(lockfile_text: str) -> Generator[FoundDependency, None, None]:
-    def extract_yarn_name(line: str) -> str:
-        """
-        "@babel/code-frame@^7.0.0", "@babel/code-frame@^7.10.4", "@babel/code-frame@^7.8.3":
-        should parse as @babel/code-frame
-
-        "@babel/types@^7.0.0", "@babel/types@^7.10.4", "@babel/types@^7.10.5", "@babel/types@^7.11.0", "@babel/types@^7.3.0", "@babel/types@^7.3.3", "@babel/types@^7.4.0", "@babel/types@^7.4.4", "@babel/types@^7.7.0", "@babel/types@^7.9.0":
-        should parse as @babel/types
-
-        lodash@4.17.18:
-        should parse as lodash
-        """
-        if '"' not in line:
-            return line.split("@")[0]
-        else:
-            first_quoted = "".join(line.split('"')[1:2])
-            parsed_name = "@".join(first_quoted.split("@")[:-1])
-            return parsed_name
+def parse_yarn(
+    lockfile_text: str, manifest_text: Optional[str]
+) -> Generator[FoundDependency, None, None]:
+    def version_sources(line: str) -> Tuple[str, List[Tuple[str, str]]]:
+        constraint_strs = line[:-1].split(",")
+        constraints = []
+        for c in constraint_strs:
+            c = c.strip().strip('"')
+            if c[0] == "@":
+                name, constraint = c[1:].split("@")
+                name = "@" + name
+            else:
+                name, constraint = c.split("@")
+            constraints.append((name, constraint))
+        return constraints[0][0], constraints
```

Comment:
Would returning Version/SemVer type here work or does it need to be strings?

Reply:
I don't think making them Versions would do anything. If "^1.0.0" appears in the manifest then "^1.0.0" will appear in the lockfile. So string equality vs semver equality will be the same.